### ISSUE TITLE
Question about training

### ISSUE THREAD

**Issue by jefferyZhan** [stevebottos/owl-vit-object-detection#3]
Hi! 
Thank you for your re-implementation of OWL. It's a nice work. I notice that there are some hyper-parameters, initialization of Linear layers and type of losses(DETR CE Loss vs Focal Loss) different from the official version. Have you trained a model achieving the similiar results to the paper?
Sincerely looking forward to your reply.

**Comment by stevebottos** [comment]
Hey @jefferyZhan, thanks for the interest. The differences you've noticed are due to the fact that this is taking the original model and repurposing it for a different task - instead of contrastively learning to encode related text + visual inputs similarly, we're now only interested in classifying/localizing objects in images. In other words this is no longer a multimodal task but rather a typical object detection task, leveraging the model's pre-existing knowledge where applicable. 

For this reason, I chose to just adopt DETR's box/class training strategy (since these are the only trainable heads in the model now, the backbone vision encoder is frozen) with a few minor tweaks to account for the fact that we have a lot more background box proposals now (DETR proposes 100 boxes while OWL-ViT proposes 768). The results for this approach aren't really comparable to the paper's because the strategy and test set are completely different (I randomly sampled 250 images for test from coco, with 2000 for training containing a small number of samples per class). Here's what I got after ten epochs:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.201
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.385
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.176
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.067
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.215
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.348
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.187
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.291
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.306
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.115
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.321
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.468
 ```
 
 So, not bad (for comparison YOLO gets like 0.64 AP@50 if I recall, after training on the full coco dataset vs my 0.385. I'll attach OWL's results at the end with a note), but definitely room for improvement. I haven't made many contributions to this repo for a while due to being pulled into some other things, but I'd love to hear any ideas that you have and feel free to open a branch of your own and contribute. 

One last note: in the paper's implementation of few shot learning, from what I can understand they focus on a single object (ie: provide ONE bounding box around an object...
![image](https://user-images.githubusercontent.com/32916895/230249045-7cc0a478-5919-40e9-9ae4-5e1d0f9695da.png)
... The content within the bounding box becomes the query. In my implementation, since this is no longer a multimodal problem, the query doesn't exist and we are going straight off the model's knowledge, since I wanted this to adapt to a more general scenario in which queries may not be available. Their results are shown below, and they're better than mine, presumably because of the fact that they're using a support query but it's entirely possible that a) my implementation can be improved or b) there may be some way to develop a bank of support queries based on the train dataset which would bring this back to a contrastive classification type of problem rather than straight up object detection.
![image](https://user-images.githubusercontent.com/32916895/230249521-c8e336b4-d9f9-4479-b951-927663796058.png)
I hope all this helps, I'm interested in hearing any suggestions for improvement!

**Comment by jefferyZhan** [comment]
Hey @stevebottos, thanks a lot for you kind reply. It has inspired me a lot, and helps me find the bugs in my code.
I try to implement this project in the original setting, including the sigmoid logits and sigmoid focal loss, which is better for imbalanced positive and negative samples. All the losses weights are set to 1.0 same as the paper. When calculating the sigmoid focal loss, it randomly choose 100 negative samples per image to calculate the classification loss. After training 22 epochs, I got the following result, a little bit higher. It might helps to improve your project too.
![image](https://user-images.githubusercontent.com/50707330/231720709-71d2c3e2-1390-4530-a768-d6234e3f14a3.png)
Also, there may be a bug in your code. The loss weight is not multiplied to the loss. As for the single shot object detection, the original implementation is better suited for this task. And in my previous experiments, the contrastive classification training does improve the AP.
I hope my reply helps, and also I'm interested in hearing your suggestions for further improvement.

**Comment by stevebottos** [comment]
@jefferyZhan Sorry for the late reply, that sounds really cool - I'll definitely have a look. What repo is yours in? If you've based off of my code could you put up a PR?

**Comment by jefferyZhan** [comment]
@stevebottos That's fine. I re-implement the whole OWL based off on Detectron2, which is quite different from your version. Sorry for cannot putting up a PR directly due to the above difference. I'll make the repo public days later after detaching the OWL from the previous codes, as the OWL part is mixed up with my other projects' codes now.

**Comment by stevebottos** [comment]
Awesome, please keep me posted. I'm very interested in exploring your implementation as well.

**Comment by jefferyZhan** [comment]
Okay! Thanks a lot!
